Animated emojis, armor crash fix, json apis

omoflop opened the pull request:
title

omoflop committed:
rendertask opacity bugfix

omoflop committed:
Armor rendering crash fix

omoflop committed:
Merge branch 'FiguraMC:1.20' into 1.20

omoflop committed:
added animated emojis

also added debug mode toggle and local cache for resource pack devs

omoflop committed:
Fixed animated emojis bug

- renamed frameTime to frametime
- added 'width' field to animated emojis as a placeholder, implementation soon :tm:

omoflop committed:
oops

Covkie commented:
If anyone has no objections ill merge this

UnlikePaladin commented:
Why is this removing the GoL easter egg without mentioning it anywhere?

UnlikePaladin commented:
If anything debug mode should be a regular config option on the settings screen and not be hidden behind a key combo that is famously used for easter eggs

UnlikePaladin commented on fabric/src/main/java/org/figuramc/figura/config/fabric/ModMenuConfig.java:
```diff
@@ -3,11 +3,12 @@
     @Override
     public ConfigScreenFactory<?> getModConfigScreenFactory() {
-        return parentScreen -> new ConfigScreen(parentScreen, FiguraMod.DEBUG_MODE);
+        return parentScreen -> new ConfigScreen(parentScreen, Configs.DEBUG_MODE.value);
```
You forgot to change this to your FiguraMod.debugModeEnabled()

UnlikePaladin commented on common/src/main/java/org/figuramc/figura/gui/screens/AbstractPanelScreen.java:
```diff
@@ -216,7 +216,8 @@ public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
         egg = egg.substring(1);
         if (EGG.equals(egg)) {
-            Minecraft.getInstance().setScreen(new GameScreen(this));
+            Configs.DEBUG_MODE.setValue("true");
```
Might as well make it a regular option on the screen, it doesn't need to be hidden behind a key combo. We already expose several options normal users don't mess with like the ip backend option.

UnlikePaladin commented on common/src/main/java/org/figuramc/figura/FiguraMod.java:
```diff
@@ -49,6 +48,13 @@ public class FiguraMod {
+        if (Configs.DEBUG_MODE == null) return false;
+        Boolean b = Configs.DEBUG_MODE.value;
+        if (b == null) return false;
+        return b;
```
you do know you could simplify this entire method to
```suggestion
   return b == true;
```
lol

UnlikePaladin reviewed (dismissed):
The rest of the changes relating to animated emojis look good to me, just these minor things.

omoflop committed:
Merge branch 'FiguraMC:1.20' into 1.20

omoflop commented:
@UnlikePaladin thanks for the feedback, will fix all that stuff 👍

omoflop committed:
fixes

+ Brought back Gol menu
+ Debug Mode & Local Cache are now not hidden on the settings screen
+ Added a emojis subcommand to the general figura command, that lists every loaded emoji
+  Cleaned up debug code

omoflop committed:
Json apis, emoji features, and motd fix

+ added parseJson(string) and toJson(any) methods to the global api
+ emojis now can optionally include a shortcuts array, which each value acts like an alias, except it doesn't need to be surrounded by `:` to be replaced
+ emojis now work in books, text displays, and the avatar menu
+ Properly hid the MOTD when there's no message, and made it scale to taller avatar info panels

+ Separated all the Emoji nested classes out into their own files
+ Refactored a few oddly named variables

omoflop committed:
Fixed armor render crash
